File: c++/leetcode/173. Binary Search Tree Iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator
{
  vector<int> sorted;
  int pos;

public:
  BSTIterator(TreeNode *root)
  {
    pos = 0;
    inorder(root);
  }

  /** @return the next smallest number */
  int next()
  {
    return sorted.at(pos++);
  }

  /** @return whether we have a next smallest number */
  bool hasNext()
  {
    return pos < sorted.size();
  }

  void inorder(TreeNode *);
};

void BSTIterator::inorder(TreeNode *root)
{
  if (!root)
    return;
  inorder(root->left);
  sorted.push_back(root->val);
  inorder(root->right);
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: javascript/leetcode/173. Binary Search Tree Iterator.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 */
var BSTIterator = function(root) {
  this._root = root;
  this._sorted = [];
  this._pos = 0;
  this.inorder(root);
};

/**
 * @return the next smallest number
 * @return {number}
 */
BSTIterator.prototype.next = function() {
  return this._sorted[this._pos++];
};

/**
 * @return whether we have a next smallest number
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function() {
  return this._pos < this._sorted.length;
};

BSTIterator.prototype.inorder = function(root) {
  if (!root) return;

  this.inorder(root.left);
  this._sorted.push(root.val);
  this.inorder(root.right);
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */
